File: 144. Binary Tree Preorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
    	vector<int> nodevec;
    	stack<TreeNode*> nodestack;
    	while(!nodestack.empty() || root != NULL) {
    	    if(root != NULL) {
    	    	nodevec.push_back(root->val);
    	    	if (root->right != NULL) {
    	    		nodestack.push(root->right);
    	    	}
	    		root = root->left;
    	    }
    	    else {
    	    	root = nodestack.top();
    	    	nodestack.pop();
    	    }
    	}
    	return nodevec;
    }
};
// recursive version
/*class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
    	vector<int> result;
		vector<int> left;
    	vector<int> right;

		if(!root)return result;    	

		result.push_back(root->val);
        left = preorderTraversal(root->left);
        right = preorderTraversal(root->right);

        if(left.size() != 0)result.insert(result.end(), left.begin(), left.end());
        if(right.size() != 0)result.insert(result.end(), right.begin(), right.end());    		    		
    	
    	return result;
    }
};*/

File: 144. Binary Tree Preorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();

        if(root == null)return result;
        stack.push(root);
        while(!stack.empty()) {
            TreeNode node = stack.pop();
            result.add(node.val);
            // since stack is LIFO, push right first
            if(node.right != null)stack.push(node.right);
            if(node.left != null)stack.push(node.left);
        }

        return result;
    }
}
